docs/available-filters/envelope-encryption.adoc

```diff
@@ -0,0 +1,155 @@
+:kms-api-javadoc: https://javadoc.io/doc/io.kroxylicious/kroxylicious-kms/latest
+
+= Envelope Encryption
+
+== What is it?
+
+A filter that provides a transparent encryption-at-rest solution for Apache Kafka.
+
+It works as follows:
+
+1. The filter intercepts the requests sent from the producing Kafka Client containing records.  The records
+   contained with the request are encrypted.  This happens before the records reach the Kafka Broker.
+2. The filter intercepts the responses sent from the Kafka Broker to the consuming Kafka Client.  The records
+   contained within the response are decrypted.  This happens before the records reach Kafka Client.
+
+The entire process is transparent from the point of view of both Kafka Client and Kafka Broker.  Neither are
+aware that the records are being encrypted.
+
+The filter encrypts the record using a symmetric encryption key.  The symmetric encryption key is held within
+a Key Management Service (KMS).
+
+Configuration is used to associate topic names with encryption keys with name encryption keys within the KMS.
+The filter uses KEK (key encryption key) selectors to decide which encryption key to use to encrypt which topic.
+
+== Prerequisites
+
+In order to use the filter, a compatible KMS (Key Management Service) must be provided.  The KMS is used to store
+the keys used to encrypt and decrypt the records sent to the topic.  The service is used at runtime. If the KMS is
+unavailable, it will become impossible to produce or consume records through the filter.  It is recommended to
+use a KMS in a highly available configuration.
+
+Currently, there is a single KMS implementation which integrates with https://www.hashicorp.com/[HashiCorp Vault&#174;].
+More KMS implementations are planned in the near future.
+
+There is also a pluggable {kms-api-javadoc}/io/kroxylicious/kms/service/KmsService.html[public API] that may be used to
+provide an alternative custom implementation.
+
+== How to use the filter
+
+There are three steps to using the filter.
+
+1. Setting up the KMS.
+2. Configuring the filter within Kroxylicious.
+3. Establishing the encryption key(s) within the KMS that will be used to encrypt the topics.
+
+These steps are described in the next sections.
+
+=== Setting up the KMS
+
+In order to set up the KMS provider ready to use with the filter, follow these provider specific steps.
+
+==== HashiCorp Vault
+
+The filter integrates with the https://developer.hashicorp.com/vault/docs/secrets/transit[HashiCorp Vault *transit
+engine*].   The transit engine is not enabled by default.  It must be
+https://developer.hashicorp.com/vault/docs/secrets/transit#setup[enabled] before it can be used with the filter.
+
+The transit engine's path must be `/transit`.
+
+The filter requires the HashiCorp Vault endpoint and a https://developer.hashicorp.com/vault/docs/concepts/tokens[token].
```

Comment:
Really need to look more closely at the requirements of the token.  At the moment, I've only been experimenting with root tokens - which is unlikely configuration as the root token will be closely guarded.